### PR TITLE
Avoid NRE and disposed exceptions when closing client twice

### PR DESCRIPTION
**Pull request by drwill-ms**
From GitHub issue #3310.

### CODE REVIEW

File: iothub/device/src/Pipeline/RetryDelegatingHandler.cs [hunk 1]
```diff
@@ -784,16 +784,17 @@ public override async Task CloseAsync(CancellationToken cancellationToken)
...
                 {
                     return;
                 }
 
                 if (Logging.IsEnabled)
                     Logging.Enter(this, cancellationToken, nameof(CloseAsync));
 
-                _handleDisconnectCts.Cancel();
-                _cancelPendingOperationsCts.Cancel();
+                _handleDisconnectCts?.Cancel();
```
Comment (drwill-ms) [inline iothub/device/src/Pipeline/RetryDelegatingHandler.cs:796]: I don't see how these get to null, but we're getting an NRE on this line when closing twice. I'd rather the client get an ObjectDisposedException than NRE, but we should be protected from even that by the change on 788.

File: e2e/test/iothub/DeviceClientE2eTests.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                // act
+                Func<Task> act = () => client.CloseAsync();
+
+                // assert
+                await act.Should().NotThrowAsync().ConfigureAwait(false);
+            }
+            finally
+            {
+                await testDevice.RemoveDeviceAsync().ConfigureAwait(false);
+                testDevice.Dispose();
```
Comment (patilsnr) [inline e2e/test/iothub/DeviceClientE2eTests.cs:39]: do we need to call dispose if already having a `using` keyword?

**Review by patilsnr — APPROVED** [review]

File: e2e/test/iothub/DeviceClientE2eTests.cs (hunk 2 above)
Comment (drwill-ms) [reply e2e/test/iothub/DeviceClientE2eTests.cs:39]: The using is on the TestDevice, which only disposes the auth, not the device client.

Comment (drwill-ms) [reply e2e/test/iothub/DeviceClientE2eTests.cs:39]: Actually, you are right. I don't need to call dispose explicitly on this one either, as there is a using on both.

**Pull request merged**